build(bazel): incrementally build aio example zips and live examples

kormide opened the pull request:
@gkalpak @devversion @josephperrott @gregmagolan 

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Examples zips and live examples are generated in bulk.

## What is the new behavior?

Example artifacts are built incrementally. The real advantage will be when these are tested incrementally, as the example e2e tests are quite expensive to run in bulk. Example tests are in progress and will come in a future PR.

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

kormide committed:
build(bazel): incrementally build aio example zips and live examples

The existing scripts were amended to work on one example at a time

kormide commented:
> Thanks for your pull request! It looks like this may be your first contribution to a Google open source project. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> View this [failed invocation](https://github.com/angular/angular/pull/46233/checks?check_run_id=6717343207) of the CLA check for more information.
> 
> For the most up to date status, view the checks section at the bottom of the pull request.

Accidentally committed with my personal email. Amended the commit to use my `aspect.dev` email.

gkalpak commented on aio/content/examples/examples.bzl:
```diff
@@ -0,0 +1,148 @@
+load("@aspect_bazel_lib//lib:copy_to_directory.bzl", "copy_to_directory")
+
+EXAMPLES = {
+    "accessibility": {"stackblitz": True, "zip": True},
```
Do we have to list all examples here? Couldn't we just detect the config files (i.e. `stackblitz.json` and `zipper.json`) as we did before?

Also, some examples (for example, `http/`) may have multiple live examples that need to be generated (by having a `stackblitz.json` and one or more `<something>.stackblitz.json` files). The new script doesn't seem to handle those.

kormide replied:
This map just makes it easier to filter down the list of examples targets to include as zips or stackblitz examples. Otherwise we'd have to specify two long lists of targets.

I was also thinking about DX for docs authors. When there's a new example, they just have to add it to this map and don't have to pass any params to the `docs_example` macro.

I wasn't aware that a single example could produce more than one live example. I'll have to take another look at this.

kormide replied:
Fixed. The map now accounts for multiple stackblitz configuration files.

kormide replied:
I also added some comments for how to use the examples map.

gkalpak commented on aio/tools/example-zipper/BUILD.bazel:
```diff
@@ -12,7 +12,7 @@ nodejs_binary(
         "@aio_npm//globby",
     ],
-    entry_point = "generateZips.mjs",
+    entry_point = "generateZip.mjs",
```
Should we also rename the `nodejs_binary` to `generate-example-zip`?

kormide replied:
Fixed

gkalpak commented on aio/tools/examples/example-boilerplate.js:
```diff
@@ -1,108 +1,61 @@
     yargs.usage('$0 <cmd> [args]')
-        .command('add', 'add the boilerplate to each example', yrgs => this.add())
-        .command('remove', 'remove the boilerplate from each example', () => this.remove())
-        .command('list-overrides', 'list all the boilerplate files that have been overridden in examples', () => this.listOverrides())
```
Could we keep the `list-overrides` command working?

kormide replied:
Sure. Just so I understand better, could you tell me what the developer workflow around `list-overrides` was? When do they call it and why?

My thinking here was that since the boilerplate and example files get combined in the output tree and not in the source tree anymore, devs won't won't really care which files are being overridden.

gkalpak replied:
`list-overrides` is useful to see if any of the overridden files need updating after a CLI update. This is explained [here][1].

[1]: https://github.com/angular/angular/blob/131d029da16b7d486cdc200f00160b9a11871fb2/aio/tools/examples/UPDATING.md#update-other-boilerplate-files

kormide replied:
Ah, I shouldn't have removed it then. I added it back.

gkalpak commented on aio/tools/examples/example-boilerplate.js:
```diff
@@ -1,108 +1,61 @@
 const fs = require('fs-extra');
```
Please also update [aio/tools/examples/README.md][1] accordingly.

[1]: https://github.com/angular/angular/blob/961d9c1d33729fbc4f5579b7d649e71b0050f997/aio/tools/examples/README.md

kormide replied:
This script shouldn't need to be called manually anymore or be a part of the developer workflow as the boilerplate is no longer generated in the source tree. I'm going to need to review all of the developer documentation before this migration is finished---do you mind if hold off on updating those docs for now?

kormide replied:
I will probably do that when I start fixing up aio's package.json to point to bazel commands.

gkalpak replied:
SGTM :+1:

gkalpak commented on aio/BUILD.bazel:
```diff
@@ -41,54 +41,28 @@ npm_package_bin(
-    },
-    output_dir = True,
-    tool = "//aio/tools/examples:example-boilerplate",
-)
```
OOC, why remove the `example-boilerplate` target?

kormide replied:
This target prepared the boilerplate for all examples in one go. Now that it's done per example, this has moved inside the `docs_example` macro.

josephperrott reviewed (commented):
I am happy with the overall direction, though I agree with @gkalpak's comments about some of the decisions and questions.

I can review as it cleans up more.

kormide commented on aio/tools/example-zipper/exampleZipper.mjs:
```diff
@@ -11,22 +11,17 @@ const __dirname = path.dirname(fileURLToPath(import.meta.url));
+    if (!fs.existsSync(configFileName)) {
+      configFileName = path.join(exampleDirName, "zipper.json");
+    }
+    this._zipExample(configFileName, exampleDirName, outputDirName);
```
Fixed

kormide commented on aio/tools/stackblitz-builder/builder.mjs:
```diff
@@ -17,22 +17,13 @@ export class StackblitzBuilder {
-      }
-    });
+    try {
+      this._buildStackblitzFrom(path.join(this.examplePath, "stackblitz.json"));
```
Fixed.

kormide commented on aio/tools/examples/create-example.js:
```diff
@@ -37,7 +37,6 @@ if (require.main === module) {
   copyExampleFiles(sourcePath, examplePath, exampleName);
```
Updated the script to copy over a templated BUILD file. It also uses `buildozer` to set the label to the example name.

kormide replied:
As for updating the EXAMPLES dict in starlark, that's a bit trickier because buildozer only supports changes to BUILD files. 

However, I added better error messaging so the bazel build will fail if you created an example but forgot to add it to the map:
```
Cannot find example foobar in the EXAMPLES map in aio/content/examples/examples.bzl. Did you forget to add an entry?
```

kormide commented on .bazelignore:
```diff
@@ -7,6 +7,7 @@ node_modules
 dist
 aio/node_modules
 aio/tools/examples/shared/node_modules
+aio/tools/examples/shared/example-scaffold
```
Don't treat the example scaffold build file as a package.

gkalpak commented on aio/tools/example-zipper/exampleZipper.mjs:
```diff
@@ -11,22 +11,25 @@ const __dirname = path.dirname(fileURLToPath(import.meta.url));
+  }
+
+  _findConfigurationFileName(exampleDirName) {
+    let configFileName = path.join(exampleDirName, "stackblitz.json");
+    if (!fs.existsSync(configFileName)) {
+      configFileName = path.join(exampleDirName, "zipper.json");
+    }
+    if (!fs.existsSync(configFileName)) {
+      throw new Error(`Missing configuration file from example directory ${exampleDirName}. Did you forget to include stackblitz.json or zipper.json?`);
+    }
+    return configFileName;
```
The logic here is a bit off. An example can have multiple zips, if it has multiple `*.stackblitz.json` files.
I believe we could keep something closer to the previous implementation (i.e. looking for all `*stackblitz.json` and `zipper.json` files, but limit it inside a single example directory) and build a ZIP for each matched file.

kormide replied:
Got it. For some reason I thought there could only be multiple live examples. I fixed this and I simplified the EXAMPLES map so that you don't have to declare the output live examples and zip names if there are multiple. The configuration files are detected in starlark now.

gkalpak commented on aio/content/examples/.gitignore:
```diff
@@ -1,46 +1,3 @@
-# boilerplate files
```
This `.gitignore` file is used and mentioned in several places. We should make sure that these changes do not affect those usages (and update docs/mentions when necessary).

- [ ] [content/guide/docs-style-guide.md](https://github.com/angular/angular/blob/131d029da16b7d486cdc200f00160b9a11871fb2/aio/content/guide/docs-style-guide.md?plain=1#L331-L347)
- [ ] [tools/examples/create-example.js](https://github.com/angular/angular/blob/131d029da16b7d486cdc200f00160b9a11871fb2/aio/tools/examples/create-example.js#L92)
- [ ] [tools/examples/create-example.spec.js](https://github.com/angular/angular/blob/131d029da16b7d486cdc200f00160b9a11871fb2/aio/tools/examples/create-example.spec.js#L85)
- [ ] [tools/examples/README.md](https://github.com/angular/angular/blob/131d029da16b7d486cdc200f00160b9a11871fb2/aio/tools/examples/README.md?plain=1#L73)
- [ ] [tools/transforms/angular-content-package/index.js](https://github.com/angular/angular/blob/131d029da16b7d486cdc200f00160b9a11871fb2/aio/tools/transforms/angular-content-package/index.js#L23)
- [ ] [tools/transforms/examples-package/processors/collect-examples.js](https://github.com/angular/angular/blob/131d029da16b7d486cdc200f00160b9a11871fb2/aio/tools/transforms/examples-package/processors/collect-examples.js#L20-L27)

kormide replied:
I took a closer look and there doesn't seem to be anything in this file that's load bearing. Now that boilerplate is not added to the source tree, and nothing is built in these example directories, each example directory should contain only the files and overrides it needs, and by default those should be included.

I suppose there's still a case for ignored examples (though there aren't any currently). Do you still want to keep this file/mechanism around or do you want me to get rid of the logic associated with it as well?

gkalpak commented on aio/content/examples/.gitignore:
```diff
@@ -1,46 +1,3 @@
-dist/
-
-
 # special
```
Many of the patterns below are no longer needed (if we get rid of the patterns above).

kormide commented on aio/tools/examples/create-example.js:
```diff
@@ -32,12 +33,13 @@ if (require.main === module) {
+  buildozer.runWithOptions([{commands: [`set name ${exampleName}`], targets: [`//aio/content/examples/${exampleName}:%docs_example`]}], {cwd: process.env.BUILD_WORKSPACE_DIRECTORY || "."});
+
   console.log(`The new "${exampleName}" example has been created.`);
-  console.log('Now run "yarn boilerplate:add" to set it up for development.');
```
Done.

kormide commented on aio/tools/stackblitz-builder/builder.mjs:
```diff
@@ -35,10 +32,20 @@ export class StackblitzBuilder {
     });
 
     if (failed) {
+      console.log(e);
```
Fixed

kormide commented on aio/tools/stackblitz-builder/builder.mjs:
```diff
@@ -35,10 +32,20 @@ export class StackblitzBuilder {
   }
 
+  _getConfigFileNames() {
+    const stackblitzPaths = path.join(this.examplePath, '**/*stackblitz.json');
```
Fixed

kormide commented on aio/content/examples/examples.bzl:
```diff
@@ -0,0 +1,168 @@
+
+    stackblitz_configs = native.glob(["*stackblitz.json"])
+
+    if EXAMPLES[name]["stackblitz"] and len(stackblitz_configs) > 0:
```
You're right that we don't need the booleans (and the map can be a list) to detect and build the zips and stackblitz examples, however the bools are necessary for the other starlark code that iterates over the map to determine the labels for all of the examples and stackblitz assets.

If we want docs authors to only have to touch one area in starlark, then the map approach is suitable. If you think it's okay that they should have to add to an examples list and add one or more labels in other files, then I can flip the map to a list. Wdyt?

kormide replied:
Three spots in `aio/content/examples/BUILD.bazel`.

gkalpak replied:
Do you mean [here][1] and [there][2] (I don't see a third spot)?

What if we got rid of the `if EXAMPLES[example]["stackblitz/zip"]` part altogether in `aio/content/examples/BUILD.bazel` (so authors would only need to update the `EXAMPLES` list in `aio/content/examples/examples.bzl`)?

[1]: https://github.com/angular/angular/blob/a2e6c94ff8206ffe1ad99f0b7b46c276be3f873d/aio/content/examples/BUILD.bazel#L15
[2]: https://github.com/angular/angular/blob/a2e6c94ff8206ffe1ad99f0b7b46c276be3f873d/aio/content/examples/BUILD.bazel#L15

kormide replied:
Ah yes, just those two. The [third usage](https://github.com/angular/angular/blob/a2e6c94ff8206ffe1ad99f0b7b46c276be3f873d/aio/content/examples/BUILD.bazel#L10) is just to make the files available for dgeni and could be done with only a list.

The `if EXAMPLES[example]["stackblitz/zip"]` isn't strictly necessary. I just thought it would be weird if `"zip": False` was set in the map but the `docs_example` macro still detects configuration for a zip and builds it anyway. Recall that the `stackblitz`/`zip` booleans just control whether the stackblitz/zip gets included in the aio build as assets via [aio/content/examples/BUILD.bazel](https://github.com/angular/angular/blob/a2e6c94ff8206ffe1ad99f0b7b46c276be3f873d/aio/content/examples/BUILD.bazel) and not whether or not to actually build those targets.

gkalpak replied:
:+1: OK, let's keep this as is for now. (We can always reconsider in the future, if necessary.)

kormide committed:
build(bazel): fix bazel aio regressions after rebase

The aio-bazel-migration branch was rebased onto main to include runfiles
support on windows.

kormide commented:
Added a commit to fix a couple of regressions caused by the recent rebase of `aio-bazel-migration` onto `main`.

gkalpak commented on aio/tests/e2e/protractor.conf.js:
```diff
@@ -4,6 +4,29 @@
 
 const { SpecReporter, StacktraceOption } = require('jasmine-spec-reporter');
 const path = require('path')
+const os = require("os");
+
+if (os.platform() === "win32") {
```
Nit 😇:

```suggestion
const os = require('os');

if (os.platform() === 'win32') {
```

gkalpak commented on aio/tests/e2e/protractor.conf.js:
```diff
@@ -4,6 +4,29 @@
+
+if (os.platform() === "win32") {
+  /*
+  For some unknown reason, the symlinked copy of chromium under runfiles won't run under
+  karma on Windows. Instead, modify the CHROME_BIN env var to point to the chrome binary
+  under external/ in the execroot.
+
+  CHROME_BIN is set to the make var $(CHROMIUM) in aio/BUILD.bazel, which points to chrome
+  under runfiles and thus starts with '../'. Because we run architect with a chdir into aio,
+  the env var is given an additional leading '../'.
+
+  First, back out of
+      bazel-out/x64_windows-fastbuild/bin/aio/e2e.bat.runfiles/angular/aio
+  Then go into
+      external/
+  and then into
+      org_chromium_chromium_windows/chrome-win
+  to cancel out the leading ../../
+*/
```
Nit: Inconsistent indentation:

```suggestion
    For some unknown reason, the symlinked copy of chromium under runfiles won't run under
    karma on Windows. Instead, modify the CHROME_BIN env var to point to the chrome binary
    under external/ in the execroot.

    CHROME_BIN is set to the make var $(CHROMIUM) in aio/BUILD.bazel, which points to chrome
    under runfiles and thus starts with '../'. Because we run architect with a chdir into aio,
    the env var is given an additional leading '../'.

    First, back out of
        bazel-out/x64_windows-fastbuild/bin/aio/e2e.bat.runfiles/angular/aio
    Then go into
        external/
    and then into
        org_chromium_chromium_windows/chrome-win
    to cancel out the leading ../../
  */
```

gkalpak commented on aio/karma.conf.js:
```diff
@@ -1,6 +1,29 @@
 // Karma configuration file, see link for more information
 // https://karma-runner.github.io/1.0/config/configuration-file.html
 
+const os = require("os");
+
+if (os.platform() === "win32") {
```
Nit 😇:

```suggestion
const os = require('os');

if (os.platform() === 'win32') {
```

gkalpak commented on aio/karma.conf.js:
```diff
@@ -1,6 +1,29 @@
+
+if (os.platform() === "win32") {
+  /*
+  For some unknown reason, the symlinked copy of chromium under runfiles won't run under
+  karma on Windows. Instead, modify the CHROME_BIN env var to point to the chrome binary
+  under external/ in the execroot.
+
+  CHROME_BIN is set to the make var $(CHROMIUM) in aio/BUILD.bazel, which points to chrome
+  under runfiles and thus starts with '../'. Because we run architect with a chdir into aio,
+  the env var is given an additional leading '../'.
+
+  First, back out of
+      bazel-out/x64_windows-fastbuild/bin/aio/test.bat.runfiles/angular/aio
+  Then go into
+      external/
+  and then into
+      org_chromium_chromium_windows/chrome-win
+  to cancel out the leading ../../
+*/
```
Nit: Inconsistent indentation:

```suggestion
    For some unknown reason, the symlinked copy of chromium under runfiles won't run under
    karma on Windows. Instead, modify the CHROME_BIN env var to point to the chrome binary
    under external/ in the execroot.

    CHROME_BIN is set to the make var $(CHROMIUM) in aio/BUILD.bazel, which points to chrome
    under runfiles and thus starts with '../'. Because we run architect with a chdir into aio,
    the env var is given an additional leading '../'.

    First, back out of
        bazel-out/x64_windows-fastbuild/bin/aio/test.bat.runfiles/angular/aio
    Then go into
        external/
    and then into
        org_chromium_chromium_windows/chrome-win
    to cancel out the leading ../../
  */
```

gkalpak commented on aio/content/examples/.gitignore:
```diff
@@ -1,108 +0,0 @@
-# boilerplate files
```
So, if this file's main(/only) purpose now is to be able to ignore an entire example by adding it here, could you please add a comment mentioning that? :pray:
(Potentially also a TODO that this functionality is unused and should be considered for removal.)

kormide replied:
I tried ignoring an example in here and it didn't play nicely with bazel. So I added a third commit to remove the ignore example feature completely and instead added a comment above the `EXAMPLES` map explaining how to ignore an example (just comment out the entry).

gkalpak reviewed (approved):
A couple last (minor) comments. Otherwise lgtm 🚀
Thx for seeing it through, @kormide ❤️ 🏋️‍♂️

Reviewed-for: global-dev-infra-approvers

kormide committed:
build(bazel): replace ignore example feature with bazel equivalent

Comment out the entry in the EXAMPLES map to ignore an example.

kormide commented on aio/tools/examples/shared/boilerplate/.gitignore:
```diff
@@ -1,3 +1,4 @@
 **/node_modules/
 **/package-lock.json
 **/yarn.lock
+**/.vscode
```
This actually should have been here before, but happened to be excluded by default by `globby`, which ignores subpaths starting with a `.`. Recall that for the bazel migration we had to set `{dot: true}` to include the bazel output tree, so that incidentally started including this folder.

josephperrott reviewed (approved):
LGTM
